Topics/Tech_Stacks/GDB_Debug.md

```diff
@@ -0,0 +1,113 @@
+
+These features make GDB an essential tool for understanding and debugging programs, whether they are executing on the same machine as GDB (native), on another machine (remote), or on a simulator. GDB is compatible with most popular UNIX and Microsoft Windows variants, as well as macOS.
+
+
+
+## How to use GDB?
+
+1. **Install GDB:**
+    1. [Linux]: 
+        Debian-based systems like Ubuntu:`sudo apt-get install gdb`
+
+        Red Hat-based systems like Fedora: `sudo dnf install gdb`
+
+    2. [macOS]: `brew install gdb`
+
+2. **Compile Your Program:**
+    Before you can debug a program with GDB, you need to compile it with debugging symbols enabled. This ensures that GDB can associate source code lines with machine code instructions.
+
+    `gcc -g -o <<program>> <<program.c>>`
+
+3. **Start GDB:**
+    `gdb program` and you will see:
+
+    ```
+    GNU gdb (Ubuntu 12.1-0ubuntu1~22.04) 12.1
+    Copyright (C) 2022 Free Software Foundation, Inc.
+    License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
+    This is free software: you are free to change and redistribute it.
+    There is NO WARRANTY, to the extent permitted by law.
+    Type "show copying" and "show warranty" for details.
+    This GDB was configured as "x86_64-linux-gnu".
+    Type "show configuration" for configuration details.
+    For bug reporting instructions, please see:
+    --Type <RET> for more, q to quit, c to continue without paging--
+    ```
+
+    and you can type `q` to quit and `c` to continue.
+
+5. **Set Breakpoints:**
+    GDB provides various ways to set breakpoints, allowing you to control where your program stops during debugging. 
+
+    1. [Line Number Breakpoint]: 
+        This is the most common type of breakpoint. It stops program execution at a specific line of source code.
+        
+        Example: `break 10`
+
+    2. [Function Name Breakpoint]: 
+        This breakpoint stops the program when it enters or exits a specified function.
+        
+        Example: `break my_function`
+
+    3. [File and Line Number Breakpoint]: 
+        This breakpoint stops the program when it enters or exits a specified function.
+        
+        Example: `break <<myfile.c>>:20`
+
+    4. [Address Breakpoint]: 
+        You can set a breakpoint at a specific memory address, which is useful for debugging assembly code or low-level programming.
+        
+        Example: `break *0x0804835c`
+
+    5. [Condition Breakpoint]: 
+         This type of breakpoint stops the program only if a specified condition is true when the breakpoint is hit.
+         
+         Example: To set a breakpoint at line 15 and break only if the variable `x` is greater than 10, use `break 15 if x > 10`.
+
+6. **Run Your Program:**
+    `run`
```

Comment:
Could add: run until a breakpoint is hit

Reply:
Thanks, I already added it